compiler/src/iree/compiler/DispatchCreation/BubbleUpExpandShapes.cpp

```diff
@@ -268,6 +268,180 @@ struct BubbleExpandThroughExtract final
+              auto dimVal =
+                  rewriter.create<tensor::DimOp>(expandOp.getLoc(), input, i);
+              expandShapeOutputDims.push_back(dimVal.getResult());
+              expandShapeOutputDimsInt.push_back(ShapedType::kDynamic);
+            } else {
+              expandShapeOutputDims.push_back(
+                  rewriter.getIndexAttr(expandedShape[k]));
+              expandShapeOutputDimsInt.push_back(expandedShape[k]);
+            }
+          } else if (concatInThisLoop &&
+                     !ShapedType::isDynamic(expandedShape[k])) {
+            nonConcatDimProduct *= expandedShape[k];
+          } else {
+            // We shouldn't reach this.
+            return rewriter.notifyMatchFailure(
+                expandOp,
+                "Expanded shape has dynamic dimension in concat reassoc"
+                "dimensions that is not the concat (ie outer) dimension");
+          }
+        }
+        // If we skipped the concat dim, we calculate it now.
+        if (k - reassoc[i].size() == newConcatDim) {
+          assert(concatInThisLoop && "concat should have been convered");
+          // Calculate the concat dim based on the product of the other
+          // dimensions.
+          if (ShapedType::isDynamic(inputShape[i])) {
+            // if input shape is dynamic, the concat dim is dynamic.
+            auto dimVal =
+                rewriter.create<tensor::DimOp>(expandOp.getLoc(), input, i);
+            expandShapeOutputDims.push_back(dimVal.getResult());
+            expandShapeOutputDimsInt.push_back(ShapedType::kDynamic);
+          } else {
+            // If input shape is static, we can calculate the concat dim.
+            expandShapeOutputDims.push_back(
+                rewriter.getIndexAttr(inputShape[i] / nonConcatDimProduct));
+            expandShapeOutputDimsInt.push_back(inputShape[i] /
+                                               nonConcatDimProduct);
+          }
+          for (int64_t j = 1; j < reassoc[i].size(); j++) {
+            // For the non-concat dimensions, we can just use the expanded
+            // shape.
+            int64_t idx = k - reassoc[i].size() + j;
+            expandShapeOutputDims.push_back(
+                rewriter.getIndexAttr(expandedShape[idx]));
+            expandShapeOutputDimsInt.push_back(expandedShape[idx]);
+          }
+        }
+      }
+      auto newType = RankedTensorType::get(expandShapeOutputDimsInt,
+                                           inputType.getElementType());
+      Value newExpand;
+      newExpand = rewriter.create<tensor::ExpandShapeOp>(
+          expandOp.getLoc(), newType, input, reassoc, expandShapeOutputDims);
+      newInputs.push_back(newExpand);
+    }
+    // Create new concat op on expanded inputs.
+    auto dimAttr = rewriter.getI64IntegerAttr(newConcatDim);
+    auto newConcat = rewriter.create<tensor::ConcatOp>(
+        concatOp.getLoc(), expandedType, dimAttr, newInputs);
```

Comment:
There is a builder that takes a `uint64_t` & doesn't need the new type

```suggestion
    auto newConcat = rewriter.create<tensor::ConcatOp>(
        concatOp.getLoc(), newConcatDim, newInputs);
```